### DISCUSSION TITLE
Clearing callbacks.

### DISCUSSION THREAD

**Discussion by Symbiomatrix** [vladmandic/sdnext#1117] [Q&A] [answered]
Hello @vladmandic, could use your guidance, please.
So I'm working on [this issue](https://github.com/hako-mikan/sd-webui-regional-prompter/issues/115), and it appears that there are two versions of our script, for txt2img and img2img. Under certain circumstances (latent mode), we add a couple callbacks through modules.script_callbacks.on_cfg_denoised + on_cfg_denoiser. However, since the scripts are unaware of one another, these callbacks are added again if the user gens both txt and img, partly the cause for the error encountered (and also wasting cycles).
I can run clear_callbacks during the extension's unload phase to fix that, but is that supposed to be handled by the specific script rather than webui? And is it safe to assume no other keys should remain on the map?

**Comment by vladmandic** [comment, marked as answer]
on a high level, there are 2 separate scriptrunner objects, one for txt2img and one for img2img (ok, there are two more, for batch processing and for api, but thats not important here). each script registers gets registered in first or second or both - depending what script says it is.
and scriptrunners execute callbacks.

typically there would be only one scriptrunner active, but there are workflows where both workflows are active at the same time - for example txt2img with latent upscale or with controlnet is basically a txt2img + img2img - its both - so you do have both scriptrunners executing their callbacks.

i'm not sure if this is what you meant with the question, i hope it helps at least a little bit.

**Reply by Symbiomatrix** [reply]
@vladmandic Thanks. It helps, but I still need some clarification.
1) Is there any situation where workflows could involve asynchronous callbacks, or are they guaranteed to alternate in the cases you mentioned?
2) I now see that clearing callback_map entirely would likely be a very bad idea, since it also contains on_ui_settings and others. If I just want to register a callback temporarily, how should I go about doing that? Register permanently and have the callback function self terminate early, or is there some other way in the current design?

**Comment by vladmandic** [comment]
re: 1 - callbacks are always synchronous. async workflows in general are rare in sd, for example creating previews for current step.
re: 2 - extensions typically do an early exit in those cases.

**Reply by Symbiomatrix** [reply]
Great, thanks. I'll do that then.
